### PR TITLE
Feature: Update checker

### PR DESCRIPTION
**Pull request by shamoon**
## Proposed change

This PR implements a little tool to check the current remote version of the app. If a newer version is found a subtle link is shown in the UI. See video below.

More detail:
- See https://github.com/paperless-ngx/paperless-ngx/pull/591#issuecomment-1086680521 . Essentially this feature is disabled by default but if it has not been explicitly set by the user a small notification is shown that the feature _can_ be enabled.
- Adds a new API endpoint `/remote_version` which checks the file `https://api.github.com/repos/paperless-ngx/paperless-ngx/releases/latest` and parses the version from that which it compares to the local version.
- Uses https://packaging.pypa.io/en/latest/version.html#packaging.version.parse (should already be in `setuptools`) for semver comparison
- Adds `PAPERLESS_ENABLE_UPDATE_CHECK` to enable / disable. Defaults to nothing which disables the feature (see [discussion below](https://github.com/paperless-ngx/paperless-ngx/pull/591#issuecomment-1086680521))
- A link is shown next to the version number in the UI, it links to our releases. Obviously this is just a heads-up for the end user.
- It should fail gracefully when no version found (e.g. http error, no internet, etc)

Though it is very little, this does contain some python stuff, so would appreciate a look over from @paperless-ngx/backend , hope I didnt butcher it too much 😅

https://user-images.githubusercontent.com/4887959/161232547-640555c4-059d-4bd5-b8b1-affda64f8448.mov

Ps. I also noticed the bootstrap icons svg was out-of-date so I updated it

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

### CODE REVIEW

**Comment by tribut** [comment]
Given that this adds a "phone-home" feature, we should add a config variable (and in my opinion, it should be disabled by default).

**Comment by shamoon** [comment]
> Given that this adds a "phone-home" feature, we should add a config variable (and in my opinion, it should be disabled by default).

I thought about this, fair enough about an option but can I ask, why off by default? I know there’s a thing about self-hosted being off-the-grid but it’s not like this is doing any kind of tracking etc. From one perspective, installing this app “phones home” in about 1000x ways (pull from npm, docker, etc). To me this seemed sort of the same, pulling from our public repo. Also there are usually potential security benefits to updating.

I don’t mean that to be contrary at all, just curious about your / everyone’s thoughts. Totally open to the discussion.

**Comment by qcasey** [comment]
I'm also curious to hear opinions on this. I'm not a fan of telemetry, but I don't think a simple GET request to one raw file on Github is worthy of the same scorn.

**Comment by shamoon** [comment]
I added a config for this, defaulting to true but of course still open for discussion on that.

File: src/documents/views.py
Comment (stumpylog) [inline src/documents/views.py:686]: I don't think this regex will match.  `\\` will match the literal `\`.

Comment (stumpylog) [reply src/documents/views.py:686]: It was probably a complaint about an invalid escape sequence.  Making it a raw string with fix that: `r"(\d+, \d+, \d+)"`

**Review by stumpylog — CHANGES_REQUESTED** [review]
One small thing for the regex.

I'm fine having this enabled by default.  If someone is conscious enough of security and any sort of request out, I would imagine they are careful about reading changelogs to find such tings.

File: src/documents/views.py
Comment (shamoon) [reply src/documents/views.py]: Ah, thanks for clarifying. Yes that was exactly the issue and python syntax is weird =)
Fixed!

**Review by stumpylog — APPROVED** [review]

**Comment by crazychatting** [comment]
I don´t think that this is a good improvement.
Sure this message is helpful, I can see the benefit for the users, but I fear this would create more technical debt/challenges and/or problems than it´s worth the lazyness to look manually now and then for a new version. Reason:
there are sure some installations completely jailed from incoming and outgoing traffic to the internet - not lan. Because obviously your document management system should never ever have the possibilities to call outwards. So in this scenario for the application it would not be possible to get the information about a new version.
I know that self-hosted Gitlab uses the Browser to make this request. But also this I would consider not really cool, because then on every usage of paperless you will tell github all the metadata about your usage with paperless... when do you use it and check for new versions, from which location are you using this etc. This is not acceptible to weight out the lazyness for checking once in a while for a new version by yourself... in my opinion at least. So this functionality should be definitely by default disabled!

I also thought that it would indeed be better to leave this functionality in the backend - but there is still scenario 1 where the app will not be able to make http requests outgoing. But for the installations where it is possible the at least this should be a timed task/cronjob in django which runs once a week/month/configurable and cache the information about new versions locally. I know, much more work, but maybe a good compromise with new features and respecting users to not unaskedly make outgoing calls.
Just my 2 cent.

**Comment by shamoon** [comment]
> Just my 2 cent.

I see your points, but I would be curious to know if thats really the majority opinion. Here's my thoughts:

1. Updating is still up to the user. Im not sure what extra "technical debt" this notification creates? Arguably less since users are perhaps more likely to get bug fixes.
2. If their install is able to communicate with the internet and they get this message they can ignore it and disable the feature. If their install is blocked then nothing happens.
3. Im not exactly sure we're telling GitHub "all the metadata". Yes, there is presumably an entry in a server log somewhere that your IP pinged for that file, that's it. If that bothers someone, again, they can disable the feature. I think the [comment above](https://github.com/paperless-ngx/paperless-ngx/pull/591#pullrequestreview-929213208) is true too, I dont think "most" users would find this so offensive.
4. Im not sure users not updating is "laziness", but whatever the reason, if we assume that new versions have security improvements then there are advantages to updating that might be worth them knowing.
5. This functionality is "in the backend", the UI does trigger it and shows it to the user. Its a good idea to check every X but that improvement could be added, to me thats not a deal-breaker as-is.
6. The reason I am suggesting enabling by default is because it sort of self-defeats the purpose of notifying the user if its disabled by default.

Again, appreciate the input. Curious to know how others feel.

**Comment by shamoon** [comment]
FYI I updated this to use `https://api.github.com/repos/paperless-ngx/paperless-ngx/releases/latest` because it occurs to me the version file could be changed but not released. Dont think that changes the discussion here too much.

File: src/documents/views.py
Comment (stumpylog) [inline src/documents/views.py:687]: It looks like the Github API comes back with JSON, so we could now `json.load` that, access tag_name and then search out the version from there.

Comment (shamoon) [reply src/documents/views.py]: Can you tell I dont do a lot go python. Cool thanks! (updated)

**Comment by gador** [comment]
I personally wouldn't mind an update checker.

For the average user, I can imagine they don't check Github on a regulatory basis to check for updates. And if there are security updates, it would be wise to inform the user of an update. 

For the security conscious user: It just checks the Github API for a new version of paperless-ngx. There is no tracking or telemetry. Even if Github would track it: It can be disabled. I would imagine a security paranoid user would also disable internet access from the paperless container/installation. 

So from me personally: I like it :+1:

File: src-ui/src/app/components/app-frame/app-frame.component.html
Comment (lippoliv) [inline src-ui/src/app/components/app-frame/app-frame.component.html:188]: The view just cares if there is an update or not. The decission wether there is an update or not doesn't matter in the view, right? 

I think it is more clean to rename `greater_than_current` to `update_available`.

```suggestion
              <div *ngIf="appRemoteVersion?.update_available" class="version-check">
```

Comment (shamoon) [reply src-ui/src/app/components/app-frame/app-frame.component.html]: Sure

File: docs/configuration.rst
Comment (qcasey) [inline docs/configuration.rst:769]:
```suggestion
    Enable (or disable) the automatic check for available updates. This works by
```

Comment (qcasey) [reply docs/configuration.rst:769]: To keep consistent with the other bools

**Comment by tribut** [comment]
Thanks for the discussion. I didn't mean to "scorn" anything or make this a bigger issue than it needs to be. However, using free software to me is to a large degree about trust. Finding out after the fact software was phoning home in some way to me is really disheartening. I understand this feature is innocuous, but as a user I would want to know about it beforehand. I know there are cultural differences to this question as well – just know that to some users, adding this without making it very obvious feels profoundly wrong.
If we decide this is something most users want and therefore should be opt-out, I suggest we 1) mention it in our `docker-compose.env` template 2) have the installer ask about it and 3) add a prominent notice about the addition to the changelog.
Do you feel this is too intrusive?

(I'm intentionally not mentioning GDPR, because I've never seen a productive legal discussion on Github :P)

**Comment by shamoon** [comment]
> Thanks for the discussion.

Well I didnt quite realize it would bring up such strong opinions! 😬

- Just as a reminder, its not phoning "home". There is no connection to the single https call made to api.github.com and this app. Github is not this app's "home" because it has no "home" because we are not a corporation with ulterior motives. Github just happens to be a publicly-accessible place where we can check the current version, thats it. Btw this website is also where we host the entire codebase and every single comment and every change, etc! Not to be glib, I just have trouble reconciling people being so concerned about this when we are discussing on this platform.
- I get that this might be a principle issue for some.
- I would be fine with making it more prominent in the installer, compose file and release notes. And of course as this is FOSS it can be always be changed / removed, etc.

**Comment by qcasey** [comment]
> I know there are cultural differences to this question as well – just know that to some users, adding this without making it very obvious feels profoundly wrong.

That's very insightful feedback, thank you.

From a technical standpoint, is it possible to make this both an .env variable *and* a checkbox in the frontend settings? It's sensible to make this opt-in if it's easy to enable (even for those users who don't read the docs and don't touch the .env files).

**Comment by shamoon** [comment]
> From a technical standpoint, is it possible to make this both an .env variable _and_ a checkbox in the frontend settings? 

Yes though that could lead to confusion, like its enabled on the front-end but not working since its disabled in the backend, etc. Not sure if theres an obvious solution to that?

**Comment by qcasey** [comment]
> > From a technical standpoint, is it possible to make this both an .env variable _and_ a checkbox in the frontend settings?
> 
> Yes though that could lead to confusion, like its enabled on the front-end but not working since its disabled in the backend, etc. Not sure if theres an obvious solution to that?

Okay, that's a good point. 

My assumption is this PR primarily helps the users copying a LinuxServer.io compose and running it without reading any docs. Or something similar to that. I don't expect that user would tweak the .env vars, use the install script, or check the changelog. I believe that same user would benefit from seeing an "Update Available" badge in the UI. 

That said, would this be best implemented as an opt-in, frontend only setting? Perhaps one that prompts the user to enable once? 

Considering the content Paperless handles, I believe it's best to not gamble trust. Especially on a relatively trivial feature.

**Comment by shamoon** [comment]
Ps. Just wanted to mention since tone can be easily misconstrued on the interwebs: though I may speak passionately about my options I have no intention of “steam-rolling” this, I defer to the group.

**Comment by qcasey** [comment]
> Ps. Just wanted to mention since tone can be easily misconstrued on the interwebs: though I may speak passionately about my options I have no intention of “steam-rolling” this, I defer to the group.

Likewise!

**Comment by shamoon** [comment]
> Considering the content Paperless handles, I believe it's best to not gamble trust. Especially on a relatively trivial feature.

Totally agree with this.

It could work as a frontend opt-in but frontend settings aren’t stored in the database! Blerg

**Comment by tribut** [comment]
Somewhat related to https://github.com/paperless-ngx/paperless-ngx/discussions/434 (I've seen this mentioned elsewhere but can't find it right now).

Not sure how linuxserver.io works exactly, but wouldn't they have to check against their registry for updates? They have lots of automation and will probably release shortly(?) after a new paperless-ngx appears, but they do also update when their own base image changes, iirc.

And to be clear: Obviously my feelings about this are no more important than anybody elses, just trying to give some input.

**Comment by qcasey** [comment]
> Not sure how linuxserver.io works exactly, but wouldn't they have to check against their registry for updates? They have lots of automation and will probably release shortly(?) after a new paperless-ngx appears, but they do also update when their own base image changes, iirc.

That's also a good point, if that URL wasn't changed by LSIO it would incorrectly alert a few days early. 

Some things like #243 and multi-user instances would benefit from a persistent frontend storage, I think that's been in our unspoken wishlist for a while now. 

Since there's dissent here and in the matrix channel, maybe this feature should be postponed? At least until a solution for easy opt-in is available?

**Comment by lippoliv** [comment]
Years ago I used to use an third party GitLab image an I've seen such behaviour. Sure, I was unhappy but in the end I switched to the official image which provided other benefits too. 

I don't want to sound rude but should paperless give third party control about shipping a feature or not? I would fear it will end up in a lot of discussions any challenges which will result in slow processes?

**Comment by qcasey** [comment]
> I don't want to sound rude but should paperless give third party control about shipping a feature or not? I would fear it will end up in a lot of discussions any challenges which will result in slow processes?

It needs to be a consideration because a seemingly large portion of paperless users are on their compose. To be clear, the LSIO URL problem won't block this feature (we can work around that), but privacy implications should.

**Comment by lippoliv** [comment]
> It needs to be a consideration because a seemingly large portion of paperless users are on their compose. To be clear, the LSIO URL problem won't block this feature (we can work around that), but privacy implications should.

Understood, privacy for sure if a good reason. (That many users do use LSIO may is a temporary situation.)

Would it be an option to release this feature disabled by default for the first time?
And maybe some release in the future could inform admins, that the default will change with the next update?

So everybody has time to prepare?

**Comment by shamoon** [comment]
I have an idea! brb…

**Comment by shamoon** [comment]
Ok everyone! I think I have a solution that should hopefully meet everyones needs / concerns. I update this PR to:

1. Feature is disabled by default.
2. If the feature has not been explicitly set as enabled or disabled, an (even more subtle) notification is displayed to the user that it hasn't been set and link to an explanation in the docs (link wont work until we merge into `main`)
3. I updated the documentation for above and also a note that e.g. linuxserver images might be behind.

Below are screenshots of all three states (obviously theres a fourth state where its enabled but no update is found in which case theres no visual indication at all). In each screenshot you can see the return from the API just for reference. Btw, the popups only show on hover so theyre pretty un-intrusive IMHO.

Again, I hope everyone is happy with this but I look forward to any feedback. Most important is to keep the good vibes here!

Screen Shot 2022-04-02 at 9 51 11 AM
Screen Shot 2022-04-02 at 10 08 47 AM
Screen Shot 2022-04-02 at 9 53 25 AM

**Comment by tribut** [comment]
I love it!

**Comment by yschroeder** [comment]
Yes, this is a good solution to the privacy concerns. However I am still unsure if the feature is required. Every feature needs maintenance and testing and might need bug fixes in the future. This might be a small feature, but many small features still lead to maintenance hell at some point (looking at owncloud and nextcloud).

So what I want to ask is:
1. Is this required to manage your documents?
2. Is it required to update this piece of software in a timely manner, because you can run into severe security issues?

I would answer it like this:
1. No, the system works and you only need to update if you encounter bugs or require a new feature.
2. This software should not run with Internet access. There will be other software in front of it (some login proxy or VPN). That other software is security critical. Paperless is not.

**Comment by shamoon** [comment]
> Yes, this is a good solution to the privacy concerns. However I am still unsure if the feature is required.

Well any 'feature' is by definition not 'required' =)

I would say cost-benefit wise there is significant potential benefit (with each release there are bug fixes that aren't necessarily pure security fixes, not to mention new features etc) and I dont think maintaining this feature is particularly challenging, the actual logic is ~30 lines of code...

**Comment by yschroeder** [comment]
Then let me rephrase: Does this PR solve a problem? Are our users facing the problem of running outdated software? Are we facing the problem of getting bug reports on outdated versions that are easily solved by updating and users being unaware they could update?

Do we implement this feature because it is possible to do it or does it solve a problem we face?

Don't get me wrong, I won't be mad if this is merged, I just want to be the conservative voice which reduces feature creep.

Let's face it: We are software developers. We can build any feature we want (god mode for software). The question is only if should we build the feature.

**Comment by shamoon** [comment]
> Are we facing the problem of getting bug reports on outdated versions that are easily solved by updating and users being unaware they could update?

Well yes, though it’s certainly not the overwhelming explanation of new issues.

Ha yes I always appreciate your conservative voice yschroder, really! I feel like there’s a philosophical debate in here too but I’ll leave that for another PR 🙂. 

Anyway, let’s see how others feel, so far my impression was that most were happy with the feature but concerned about the privacy etc aspects which hopefully are now addressed for most.

**Comment by s-martin** [comment]
I think from a users perspective it makes sense, because it helps users who don't check GitHub regularly.

The proposed solution is a good tradeoff between privacy and comfort.

**Comment by lippoliv** [comment]
> Then let me rephrase: Does this PR solve a problem? Are our users facing the problem of running outdated software? Are we facing the problem of getting bug reports on outdated versions that are easily solved by updating and users being unaware they could update?

If I put me in the Position of a User, am used to have notifications of updates in many many applications. Most use dialogs, which I think is "a little bit too much". I really like @shamoon's solution, it's decent.

So I would say, if paperless would not support that, probably at some point there will be a feature request by some user anyways. 

> I just want to be the conservative voice which reduces feature creep.

This is probably a good idea, esp. since GitHub API is leveraged here. They probably change their API at some point and this hardly can be covered by tests I guess. Everything else can be covered, so maintenance level should be low.

File: src-ui/src/app/components/app-frame/app-frame.component.html
Comment (crazychatting) [inline src-ui/src/app/components/app-frame/app-frame.component.html:198]: can the color of the icon be less intrusive for the state `updateCheckNotSet`? maybe the same as the text-color?

File: src/paperless/settings.py
Comment (crazychatting) [inline src/paperless/settings.py:572]: I understood this that if the value is anything else than `false` or [unset] the feature will be enabled, correct? wouldn´t it be better to do this vice-versa - ony enable it if the value is explicit `true`?

Comment (tribut) [reply src/paperless/settings.py:572]: No, the default on line 572 is misleading, because an unset `PAPERLESS_ENABLE_UPDATE_CHECK` would already be caught in line 570. Unless the environment changes mid-script, this value is never used. `__get_boolean` itself errs on the side of "false".

**Comment by shamoon** [comment]
> I wonder if it should be possible to enable this from the UI. Could we store this setting in the database instead and have it enabled with just one click?

I agree but that is true for many UI settings so I'd rather not include that here. At the moment this is essentially a 'workaround' for this fact. But if / when we are ready to pursue making some / many / all of the frontend settings stored in the DB I would love to help with that PR. TBH I could probably figure it out but python/Django are not my strong suit whereas others here are very comfortable with them so seems like would be better to collaborate. If anyone is game, just holler!

**Comment by Chumper** [comment]
I am currently checking every week if there is a new update for paperless, I would not need to do that if the feature would be present.

I am from germany, so privacy is a big factor here. 
However I am fine with is being disabled by default and that it needs to be explicitly enabled.
Also the term phoning home does really fit in this case, because as mentioned above there is not data sent to anywhere, except that there is a call to github. 

I however would argue for checking the version from the browser, not from the server. 
As some folks mentioned, the server could potentially sit behind a firewall and have no chance to call the API.
Calling from the browser also hides any information one could get from the system.
For GitHub it looks like a "manual" call from a user that will probably visit GitHub anyway.

**Comment by shamoon** [comment]
The reasons for checking from the backend are technical, pretty much due to CORS.

Glad to know another privacy-conscious user is OK with this.

If I'm reading the room I think this should be OK to merge now. Even if not everyone wants to use this it seems the compromise has been acceptable. Not in a rush but will plan to do that soon. Of course discussion is always open, appreciate everyones thoughts.

**Review by qcasey — APPROVED** [review]
Agreed, the value to maintenance burden ratio makes sense here IMO.

**Pull request merged**
